### PR TITLE
Events

### PR DESCRIPTION
**Pull request by GrandSchtroumpf**
Use the [events library](https://www.npmjs.com/package/events).
Update documentation.
Update examples.
Update Tests.

### CODE REVIEW

File: examples/modules/txlistener.module.ts [hunk 1]
```diff
@@ -32,9 +35,9 @@ export class TxlistenerApi extends API<Txlistener> implements Txlistener {
 // External class
 export class TxEmitter {
```
Comment (yann300) [inline examples/modules/txlistener.module.ts]: In the current implementation (remix IDE code), I am not sure we need to have a `TxEmitter` . the current `TxListener` is listening either on new blocks or in a ethereumjs-vm event but it does not need to register directly to an emitter.
but that's a detail, we can merge this API def with the current implementation when it's time

File: src/app-manager.ts [hunk 2]
```diff
@@ -64,26 +72,21 @@ export class AppManager {
...
+  private activateApi<M extends Api>(json: ModuleProfile<M>, api: API<M>) {
     const events = json.events || []
     events.forEach(event => {
-      if (event in api) {
-        api[event].on((value: any) => {
-          this.broadcast(api.type, event, value)
-        })
-      }
+      if (!api.events) return
+      api.events.on(event, (value: any) => this.broadcast(api.type, event, value))
```
Comment (yann300) [inline src/app-manager.ts]: does we need the template here on the call to `broadcast`

File: src/app-manager.ts [hunk 3]
```diff
@@ -130,12 +134,12 @@ export class AppManager {
 
     const methods = json.methods || []
     methods.forEach(key => {
-      if (key in api) delete this[api.type][key]
+      if (key in api) delete this[api.type as string][key]
```
Comment (yann300) [inline src/app-manager.ts]: why do we need `as string` here? `type` is not already enforced to be a string as it is declared in the interface `API`?

File: examples/modules/txlistener.module.ts (hunk 1 above)
Comment (GrandSchtroumpf) [reply examples/modules/txlistener.module.ts]: It's just an example of how a module can deal with an external class. In the end the implementation won't look exactly like that.

File: src/app-manager.ts (hunk 2 above)
Comment (GrandSchtroumpf) [reply src/app-manager.ts]: What do you mean by "template" ?

File: src/app-manager.ts (hunk 3 above)
Comment (GrandSchtroumpf) [reply src/app-manager.ts]: In this case, it's kind of a mismatch between the generic Api "M" in this method and what the `this.calls` method accept with the generic IAppManager "T" in the class. Removing the `as string` create a typescript warning. It's a quite fix to remove a Typescript warning, it won't impact the end Javascript code. But yes, I should improve that in time.

File: src/types.ts [hunk 4]
```diff
@@ -1,24 +1,42 @@
...
-  events?: (Extract<keyof T, string>)[]
+  methods?: ExtractKey<T, Function>[]
+  events?: (keyof T['events'])[]
 }
 
-export interface PluginProfile {
-  type: string
-  methods?: string[]
-  events?: string[]
+export interface PluginProfile<T extends Api = any> extends ModuleProfile<T> {
```
Comment (yann300) [inline src/types.ts]: `PluginProfile` refers to a plugin JSON declaration right?
why removing `type`, `methods`, and `events`?

Comment (GrandSchtroumpf) [reply src/types.ts]: I'm not, it's extending the `ModuleProfile` instead which has the `type`, `methods` and `events`.

**Pull request merged**